planner/core/rule_count_star_rewriter.go

```diff
@@ -0,0 +1,101 @@
+package core
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/parser/mysql"
+)
+
+/**
+The countStarRewriter is used to rewrite count(*) -> count(not null column)
+
+Pattern:
+LogcialAggregation
+       |
+   DataSource
+
+Optimize:
+Table
+  <k1 bool not null, k2 int null, k3 bigint not null>
+
+Case1 there are columns exists in datasource
+Query: select count(*) from table where k3=1
+CountStarRewriterRule: pick the narrowest not null column exists in datasource
+Rewritten Query: select count(k3) from table where k3=1
+
+Case2 there is no columns exists in datasource
+Query: select count(*) from table
+ColumnPruningRule: pick k1 as the narrowest not null column from origin table @Function.preferNotNullColumnFromTable
+                   datasource.columns: k1
+CountStarRewriterRule: rewrite count(*) -> count(k1)
+Rewritten Query: select count(k1) from table
+
+Attention:
+Since count(*) is directly translated into count(1) during grammar parsing,
+the rewritten pattern actually matches count(1)
+*/
+
+type countStarRewriter struct {
+}
+
+func (c *countStarRewriter) optimize(ctx context.Context, p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	return c.countStarRewriter(p, opt)
+}
+
+func (c *countStarRewriter) countStarRewriter(p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	// match pattern agg(count(*)) -> datasource
+	if agg, ok := p.(*LogicalAggregation); ok {
+		if agg.GroupByItems == nil || len(agg.GroupByItems) == 0 {
+			if dataSource, ok := agg.Children()[0].(*DataSource); ok {
+				for _, aggFunc := range agg.AggFuncs {
+					if aggFunc.Name == "count" && len(aggFunc.Args) == 1 && !aggFunc.HasDistinct {
+						if constExpr, ok := aggFunc.Args[0].(*expression.Constant); ok {
+							if constExpr.Value.GetInt64() == 1 {
```

Comment:
I think as long as the constant is not null, we can process it.
we can do it for count(1), count(100), count('xxx'), count(1.2323), but not `count(null)`, right?

Reply:
You are right. Maybe the rule should be named ```rule_count_constant_rewriter.go``` is better ?

Reply:
The name is ok, we don't need to change it.